pipeline-model-definition/src/main/resources/org/jenkinsci/plugins/pipeline/modeldefinition/ModelInterpreter.groovy

```diff
@@ -365,77 +380,74 @@ public class ModelInterpreter implements Serializable {
                             setUpDelegate(postClosures.get(ni))
                         }
                     } catch (Exception e) {
-                        script.echo "Error in stage post: ${e.getMessage()}"
                         script.getProperty("currentBuild").result = Result.FAILURE
-                        if (firstError == null) {
-                            firstError = e
+                        if (stageError == null) {
+                            stageError = e
                         }
                     }
                 }
             }
         }
 
-        return firstError
+        if (stageError != null) {
+            throw stageError
+        }
     }
 
     /**
      * Executes the notifications for this build.
      * @param root The root context we're running in.
-     * @param firstError The current error state
-     * @return The updated current error state
      */
-    def executeNotifications(Root root, Throwable firstError) {
-        try {
-            // And finally, run the notifications.
-            List<Closure> notificationClosures = root.satisfiedNotifications(script.getProperty("currentBuild"))
+    def executeNotifications(Root root) throws Throwable {
+        Throwable stageError
 
-            catchRequiredContextForNode(root.agent, true) {
-                if (notificationClosures.size() > 0) {
-                    script.stage("Notifications") {
+        // And finally, run the notifications.
+        List<Closure> notificationClosures = root.satisfiedNotifications(script.getProperty("currentBuild"))
+        if (notificationClosures.size() > 0) {
+            try {
+                script.stage("Notifications") {
+                    catchRequiredContextForNode(root.agent, true) {
                         for (int i = 0; i < notificationClosures.size(); i++) {
                             setUpDelegate(notificationClosures.get(i))
                         }
                     }
                 }
-            }
-        } catch (Exception e) {
-            script.echo "Error in notifications execution: ${e.getMessage()}"
-            script.getProperty("currentBuild").result = Result.FAILURE
-            if (firstError == null) {
-                firstError = e
+            } catch (Exception e) {
+                script.getProperty("currentBuild").result = Result.FAILURE
```

Comment:
Be really really cautious with doing this.  I don't see an error here but explicitly setting status can lead to some... odd behaviors.